src/java/org/apache/ivy/core/cache/DefaultRepositoryCacheManager.java

```diff
@@ -1215,26 +1217,43 @@ public ResolvedResource resolve(Artifact artifact) {
                     for (int i = 0; i < confs.length; i++) {
                         Artifact[] arts = md.getArtifacts(confs[i]);
                         for (int j = 0; j < arts.length; j++) {
-                            Artifact transformedArtifact = NameSpaceHelper.transform(
-                                arts[j], options.getNamespace().getToSystemTransformer());
-                            ArtifactOrigin origin = getSavedArtifactOrigin(
-                                transformedArtifact);
-                            File artFile = getArchiveFileInCache(
-                                transformedArtifact, origin, false);
-                            if (artFile.exists()) {
-                                Message.debug("deleting " + artFile);
-                                if (!artFile.delete()) {
-                                    // Old artifacts couldn't get deleted!
-                                    // Restore the original ivy file so the next time we
-                                    // resolve the old artifacts are deleted again
-                                    backupDownloader.restore();
-                                    Message.error("Couldn't delete outdated artifact from cache: " + artFile);
-                                    return null;
-                                }
-                            }
-                            removeSavedArtifactOrigin(transformedArtifact);
+                            if (!prepAndDeleteArtifact(arts[j], options, backupDownloader)) return null;
                         }
                     }
+
+                    // sbt change, driven by https://github.com/sbt/sbt/issues/1750
+                    //
+                    // Typically Ivy behaviour, in the parsing above
+                    // (with ModuleDescriptorParser & PomModuleDescriptorBuilder),
+                    // involves making remote calls to locate sources and javadoc jars on the remote repository
+                    // and storing them in the ModuleDescriptor.
+                    //
+                    // However given a number of dependencies and resolver this suffers in performance,
+                    // particurlarly Ivy's lack of caching for not finding artifacts or IO failures.
+                    //
+                    // So sbt overrides this behaviour by failing fast (returning null).
+                    // See IvySbt.hasImplicitClassifier.
+                    //
+                    // This however has the side-effect that sources and javadoc jars aren't deleted here
+                    // so that they can be re-downloaded.
+                    //
+                    // So we keep the performance overrides and just delete those jars directly below.
```

Comment:
Ok, so not to be annoying, BUT...

1. This implies we have the performance overrides in this project, which we may actually want to do.
2. IIRC there's an open ticket and PR to disable downloading source/javadoc to the Ivy main REPO, so we're not out of line just putting all that functionality in here.

I'd prefer if we have ONE of the hackeries in this repo that we just keep them BOTH together so they can reference each other.